File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.py
import heapq

def solution(book_time):
    answer = 0
    # 퇴실시간 ~ 퇴실시간+10분 사용못함 이때까지
    # 최소 객실의 수
    
    book_time.sort (key = lambda x :(x[0]))
    
    change_time = []
    
    for start, end  in book_time :
        start_n = int (start.split(":")[0])*60 
        start_m = int (start.split(":")[1]) 
        
        s = start_n +start_m
        
        end_h = int (end.split(":")[0])*60
        end_m = int (end.split(":")[1])
        
        e = end_h +end_m+10
        
        change_time.append((s,e))
        
    rooms  = []
    
    for s,e in change_time :
        if rooms and rooms[0] <= s:
            heapq.heappop(rooms)
        else :
            answer+=1
        heapq.heappush(rooms,e)
     

    return answer

File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.java
import java.util.*;

class Solution {
    public int solution(String[][] book_time) {
        int answer = 0;
        
        int[][] times = new int [book_time.length][2];
        
        int idx= 0;
        
        for (String [] target : book_time) {
            int start_h = Integer.parseInt(target[0].substring(0,2)) * 60;
            int start_m = Integer.parseInt(target[0].substring(3,5));
            
            int end_h = Integer.parseInt(target[1].substring(0,2))*60;
            int end_m = Integer.parseInt(target[1].substring(3,5)) +10;
            
            int s = start_h+start_m;
            int e = end_h +end_m;
            
            times[idx][0] = s;
            times[idx][1] =e;
            idx +=1;
            
        }
        
        Arrays.sort (times, (a,b) -> a[0]-b[0]);
        
//         for (int[] a : times) {
//              System.out.println(Arrays.toString(a));
//         }
        
        PriorityQueue <Integer> pq = new PriorityQueue<>();
        
        for (int [] target : times ) {
            int start = target[0];
            int end = target[1];
            
            if (!pq.isEmpty() && pq.peek() <= start) {
                pq.poll(); 
            }else {
                answer+=1;
            }
            pq.offer (end);

        }
       
        return answer;
    }
}